File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])

        def dfs(r,c):
            if r<0 or r>=ROWS or c<0 or c>=COLS or grid[r][c] == 0:
                return 0
            if grid[r][c]==1:
                grid[r][c]=0

            return (1 + dfs(r+1, c) + dfs(r-1, c) + dfs(r, c+1) + dfs(r, c-1))

        area = 0
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c]==1:
                    area = max(area,dfs(r,c))
        
        return area

File: Data Structures & Algorithms/interleaving-string/submission-1.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        return dfs(s1, s2, s3, 0, 0, 0);
    }

    bool dfs(string s1, string s2, string s3 , int i1 ,int i2, int i3){
        if(i3==s3.size()){
            return true;
        }

        if(s1[i1]==s3[i3] && s2[i2]==s3[i3]){
            return dfs(s1, s2, s3, i1+1, i2, i3+1) || dfs(s1, s2, s3, i1, i2+1, i3+1);
        }

        else if(s1[i1]==s3[i3]){
            return dfs(s1, s2, s3, i1+1, i2, i3+1);
        }
        else if(s2[i2]==s3[i3]){
            return dfs(s1, s2, s3, i1, i2+1, i3+1);
        }

        return false;

    }
};
